library/core/src/iter/traits/iterator.rs

```diff
@@ -3860,6 +3839,46 @@ pub trait Iterator {
     }
 }
 
+/// Compares two iterators element-wise using the given function.
+///
+/// If `ControlFlow::CONTINUE` is returned from the function, the comparison moves on to the next
+/// elements of both iterators. Returning `ControlFlow::Break(x)` short-circuits the iteration and
+/// returns `ControlFlow::Break(x)`. If one of the iterators runs out of elements,
+/// `ControlFlow::Continue(ord)` is returned where `ord` is the result of comparing the lengths of
+/// the iterators.
+///
+/// Isolates the logic shared by ['cmp_by'](Iterator::cmp_by),
+/// ['partial_cmp_by'](Iterator::partial_cmp_by), and ['eq_by'](Iterator::eq_by).
+#[inline]
+fn iter_compare<A, B, F, T>(mut a: A, mut b: B, f: F) -> ControlFlow<T, Ordering>
+where
+    A: Iterator,
+    B: Iterator,
+    F: FnMut(A::Item, B::Item) -> ControlFlow<T>,
+{
+    #[inline]
+    fn compare<'a, B, X, T>(
+        b: &'a mut B,
+        mut f: impl FnMut(X, B::Item) -> ControlFlow<T> + 'a,
+    ) -> impl FnMut(X) -> ControlFlow<ControlFlow<T, Ordering>> + 'a
+    where
+        B: Iterator,
+    {
+        move |x| match b.next() {
+            None => ControlFlow::Break(ControlFlow::Continue(Ordering::Greater)),
+            Some(y) => f(x, y).map_break(ControlFlow::Break),
+        }
+    }
+
+    match a.try_for_each(compare(&mut b, f)) {
+        ControlFlow::Continue(()) => ControlFlow::Continue(match b.next() {
```

Comment:
This is one of the core problems with inner iteration, in a way.  You can't naturally loop over two things at once with it, so you're stuck prioritizing one of them and single-stepping the other.  (Now, single-stepping can be spelled `.try_for_each(ControlFlow::break).break_value()`, so it's not like it's a blocker to writing stuff, but this necessity of single-stepping is one of the reasons iterators changed way back in the day to being based around `next`, IIRC.)

The big thing I'm noticing in this PR is that these functions are over two *owned* iterators.  That means (unlike if it was `&mut self`, say) that I think this could use `zip`?  Then this would be able to take advantage of the `zip` optimizations, and would still have a good fallback implementation without needing to repeat that logic here.